Rcl issues 469 test arguments

y-okumura-isp opened the pull request:
Related with #469.

This is PR about problems in test_arguments found by using asan.

- rcl_arguments
  - memory leak and initializtion problem about external_log_config_file handling
- rcl_yaml_param_parser
  - memory leak

There are still 2 problems(seems complicated).

clalancette commented on rcl_yaml_param_parser/src/parser.c:
```diff
@@ -1345,7 +1345,9 @@ static rcutils_ret_t parse_key(
-          ret = rem_name_from_ns(ns_tracker, NS_TYPE_NODE, allocator);
+	  allocator.deallocate(node_name_ns, allocator.state);
+
+	  ret = rem_name_from_ns(ns_tracker, NS_TYPE_NODE, allocator);
```
There's some weird indentation happening here; probably tabs instead of spaces.

y-okumura-isp replied:
Thanks. I fix it (bca8dad79f057385a5756e0ee28a3d45ce447478)

clalancette commented:
> Can you please provide the output that asan was complaining about for these problems? It will make it easier for us to understand if this is actually fixing the issue. Thanks.

Ah, sorry, I didn't follow the link.  It's in the attached issue.  I'll continue reviewing.

clalancette commented on rcl/src/rcl/arguments.c:
```diff
@@ -958,6 +958,7 @@ rcl_arguments_copy(
   args_out->impl->parameter_overrides = NULL;
   args_out->impl->parameter_files = NULL;
   args_out->impl->num_param_files_args = 0;
+  args_out->impl->external_log_config_file = NULL;
```
So this is basically duplicating the same code that is at https://github.com/ros2/rcl/blob/7537f8304493e9b9568748b5601f5c1b7fa463f0/rcl/src/rcl/arguments.c#L330 (which is why it was probably missed).

I'm going to suggest that instead of adding this line, we add in a method that is something like `rcl_allocate_and_zero`, which will both do the allocation and setting all of these to default values.  That way, we can't forget something like this in the future.

y-okumura-isp replied:
I have the same opinion and do it (183ad826d14e7257d72913bfee36b6360602176d).

Some variables are set in rcl_parse_arguments() but not in rcl_arguments_copy() such as
- log_level
- log_stdout_disabled
- log_rosout_disabled
- log_ext_lib_disabled

Do you know they are only forgotten or there is some reason?
I left it as is because I don't understand rcl_arguments_copy use case in detail.

clalancette replied:
>     * 
> 
> 
> Do you know they are only forgotten or there is some reason?
> I left it as is because I don't understand rcl_arguments_copy use case in detail.

It's not entirely clear to me, but I believe we want to copy them as well.

y-okumura-isp replied:
@clalancette Thank you for comment. I fixed.

y-okumura-isp commented:
@clalancette

Thank you for review.

> Can you please provide the output that asan was complaining about for these problems?

Here is the asan output related with 0c37b206eca59a60392c53718d7182bb6039cfe8 (about external_log_config_file).

```
    #0 0x7fe06c8c4b50 in __interceptor_malloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdeb50)
    #1 0x7fe06c0fc885 in __default_allocate /ros2_asan/src/ros2/rcutils/src/allocator.c:35
    #2 0x7fe06c10f57a in rcutils_strndup /ros2_asan/src/ros2/rcutils/src/strdup.c:42
    #3 0x7fe06c10f4f9 in rcutils_strdup /ros2_asan/src/ros2/rcutils/src/strdup.c:33
    #4 0x7fe06c56d854 in _rcl_parse_external_log_config_file /ros2_asan/src/ros2/rcl/rcl/src/rcl/arguments.c:1959
    #5 0x7fe06c55e41a in rcl_parse_arguments /ros2_asan/src/ros2/rcl/rcl/src/rcl/arguments.c:518
    #6 0x55d5ce713cfe in are_known_ros_args(std::vector<char const*, std::allocator<char const*> >) /ros2_asan/src/ros2/rcl/rcl/test/rcl/test_arguments.cpp:116
    #7 0x55d5ce71cb97 in TestArgumentsFixture__rmw_fastrtps_cpp_check_known_vs_unknown_args_Test::TestBody() /ros2_asan/src/ros2/rcl/rcl/test/rcl/test_arguments.cpp:168
    (snip gtest traces)
```


And for 142ef3532b9e4fca20d0d3fcc88bd970d8260bab (about node_name_ns deallocate).

```
   #1 0x7fcbe6fc9885 in __default_allocate /ros2_asan/src/ros2/rcutils/src/allocator.c:35
   #2 0x7fcbe6fdc57a in rcutils_strndup /ros2_asan/src/ros2/rcutils/src/strdup.c:42
   #3 0x7fcbe6fdc4f9 in rcutils_strdup /ros2_asan/src/ros2/rcutils/src/strdup.c:33
   #4 0x7fcbe71fdcdb in parse_key /ros2_asan/src/ros2/rcl/rcl_yaml_param_parser/src/parser.c:1337
   #5 0x7fcbe71ff055 in parse_file_events /ros2_asan/src/ros2/rcl/rcl_yaml_param_parser/src/parser.c:1481
   #6 0x7fcbe7200751 in rcl_parse_yaml_file /ros2_asan/src/ros2/rcl/rcl_yaml_param_parser/src/parser.c:1694
   #7 0x7fcbe743a706 in _rcl_parse_param_file /ros2_asan/src/ros2/rcl/rcl/src/rcl/arguments.c:1919
   #8 0x7fcbe7429bc1 in rcl_parse_arguments /ros2_asan/src/ros2/rcl/rcl/src/rcl/arguments.c:455
   #9 0x5610d1924cfe in are_known_ros_args(std::vector<char const*, std::allocator<char const*> >) /ros2_asan/src/ros2/rcl/rcl/test/rcl/test_arguments.cpp:116
   #10 0x5610d192c345 in TestArgumentsFixture__rmw_fastrtps_cpp_check_known_vs_unknown_args_Test::TestBody() /ros2_asan/src/ros2/rcl/rcl/test/rcl/test_arguments.cpp:158
   (snip gtest traces)
```

ros-discourse commented:
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/introducing-ros2-sanitizer-report-and-analysis/9287/6

clalancette commented on rcl/src/rcl/arguments.c:
```diff
@@ -941,23 +936,12 @@ rcl_arguments_copy(
-  args_out->impl->parameter_overrides = NULL;
-  args_out->impl->parameter_files = NULL;
-  args_out->impl->num_param_files_args = 0;
+  do {
```
Any particular reason for this `do{ ... } while (false)` loop?

y-okumura-isp replied:
Thank you for comment.
I wanted to narrow the scope of "ret".

If "ret" is defined in the function scope, it seems to be rcl_arguemnts_copy return value.
But there are lines which directly return values, ex `return RCL_RET_BAD_ALLOC;` or `return return RCL_RET_OK;`. I feel "why not return ret!?".

As I want to use "ret" very temporarily, so I use `do{ ..} while(false)`

How do you think these ideas?

(1) simply use "ret" (line 991 is also affected)
(2) change variable name such as "_ret".
(3) bellow

```
if (RCL_RET_OK != _rcl_allocate_initialized_arguments_impl(args_out, &allocator)) {
  return ret;
}
```

clalancette replied:
The problem with 3) is that you won't have the correct value to return.  You could do assignment in the `if` statement, but I generally find that hard to read.

In this case, I think we just have to accept that this is a limitation of C and just leave `ret` as function scoped.  Since we are reducing the complexity of this function anyway, I don't think it will be hard to read.

clalancette commented on rcl_yaml_param_parser/src/parser.c:
```diff
@@ -1345,7 +1345,9 @@ static rcutils_ret_t parse_key(
-          ret = rem_name_from_ns(ns_tracker, NS_TYPE_NODE, allocator);
+          allocator.deallocate(node_name_ns, allocator.state);
+
+	  ret = rem_name_from_ns(ns_tracker, NS_TYPE_NODE, allocator);
```
Indentation seems weird here again.

clalancette commented on rcl_yaml_param_parser/src/parser.c:
```diff
@@ -1345,7 +1345,9 @@ static rcutils_ret_t parse_key(
           }
 
-          ret = rem_name_from_ns(ns_tracker, NS_TYPE_NODE, allocator);
+          allocator.deallocate(node_name_ns, allocator.state);
```
Isn't there another leak up above if `find_node` returns something other than `RCUTILS_RET_OK`?

y-okumura-isp replied:
How about `deallocate` right after find_node?

```
ret = find_node(node_name_ns, params_st, node_idx);
allocator.deallocate(node_name_ns, allocator.state);
if (RCUTILS_RET_OK != ret) {
  break;
}
```

clalancette replied:
Yeah, that should fix it.

y-okumura-isp commented:
A supplement to the PR above.

I fixed above change requests.
Also I fixed my old change(arguments.c:1105-1106) because ament_uncrustify claims.
So I rewrote it such as https://github.com/ros2/rcl/blob/master/rcl/src/rcl/arguments.c#L1007.

```
// ament_uncrustify claim
--- src/ros2/rcl/rcl/src/rcl/arguments.c
+++ src/ros2/rcl/rcl/src/rcl/arguments.c.uncrustify
@@ -1104 +1104 @@
-                                       args->impl->allocator.state);
+        args->impl->allocator.state);


// my changes at arguments.c:1105-1106
-      args->impl->allocator.deallocate(args->impl->external_log_config_file,
-                                       args->impl->allocator.state);
+      args->impl->allocator.deallocate(
+        args->impl->external_log_config_file, args->impl->allocator.state);
```

Sorry for not accustomed to ROS 2 coding style, I'm little confused.

I thought [ROS 2 follows Google C++ coding style](https://index.ros.org/doc/ros2/Contributing/Code-Style-Language-Versions/#id2), and in https://google.github.io/styleguide/cppguide.html#Function_Calls,
subsequent lines should be aligned with the first argument.
And my editor with google-c-style alignes as such.
But ament_uncrustify claims second argument should be aligned more shallow...

Anyway I hope it's right change.

clalancette commented:
> I thought [ROS 2 follows Google C++ coding style](https://index.ros.org/doc/ros2/Contributing/Code-Style-Language-Versions/#id2)

We do use the Google C++ coding style, mostly.  The problem is that uncrustify doesn't totally support that style, so there are sometimes differences between what uncrustify wants and what the Google style would suggest.  As long as uncrustify passes, that is the definitive source for us.

clalancette commented:
CI:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=9551)](http://ci.ros2.org/job/ci_linux/9551/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=5254)](http://ci.ros2.org/job/ci_linux-aarch64/5254/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=7799)](http://ci.ros2.org/job/ci_osx/7799/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=9474)](http://ci.ros2.org/job/ci_windows/9474/)

clalancette commented:
All of the test failures seen here were in other builds, so they weren't caused by this PR.  Thus I'm going to go ahead and merge this.  Thanks for the contribution.
